File: db/rust/src/entity/company.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// 取引先マスタの構造体
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, FromRow)]
pub struct Company {
    #[sqlx(rename = "取引先コード")]
    pub comp_code: String,
    #[sqlx(rename = "取引先名")]
    pub name: String,
    #[sqlx(rename = "取引先名カナ")]
    pub kana: Option<String>,
    #[sqlx(rename = "仕入先区分")]
    pub sup_type: Option<i32>,
    #[sqlx(rename = "郵便番号")]
    pub zip_code: Option<String>,
    #[sqlx(rename = "都道府県")]
    pub state: Option<String>,
    #[sqlx(rename = "住所１")]
    pub address1: Option<String>,
    #[sqlx(rename = "住所２")]
    pub address2: Option<String>,
    #[sqlx(rename = "取引禁止フラグ")]
    pub no_sales_flg: Option<i32>,
    #[sqlx(rename = "雑区分")]
    pub wide_use_type: Option<i32>,
    #[sqlx(rename = "取引先グループコード")]
    pub comp_group_code: String,
    #[sqlx(rename = "与信限度額")]
    pub max_credit: Option<i32>,
    #[sqlx(rename = "与信一時増加枠")]
    pub temp_credit_up: Option<i32>,
    #[sqlx(rename = "作成日時")]
    pub create_date: NaiveDateTime,
    #[sqlx(rename = "作成者名")]
    pub creator: Option<String>,
    #[sqlx(rename = "更新日時")]
    pub update_date: NaiveDateTime,
    #[sqlx(rename = "更新者名")]
    pub updater: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::NaiveDate;

    fn create_test_company() -> Company {
        Company {
            comp_code: "COMP001".to_string(),
            name: "テスト株式会社".to_string(),
            kana: Some("テストカブシキガイシャ".to_string()),
            sup_type: Some(1),
            zip_code: Some("100-0001".to_string()),
            state: Some("東京都".to_string()),
            address1: Some("千代田区".to_string()),
            address2: Some("1-1-1".to_string()),
            no_sales_flg: Some(0),
            wide_use_type: Some(0),
            comp_group_code: "GRP001".to_string(),
            max_credit: Some(10000000),
            temp_credit_up: Some(0),
            create_date: NaiveDate::from_ymd_opt(2021, 1, 1).unwrap().and_hms_opt(0, 0, 0).unwrap(),
            creator: Some("admin".to_string()),
            update_date: NaiveDate::from_ymd_opt(2021, 1, 1).unwrap().and_hms_opt(0, 0, 0).unwrap(),
            updater: Some("admin".to_string()),
        }
    }

    #[test]
    fn test_company_creation() {
        let company = create_test_company();
        assert_eq!(company.comp_code, "COMP001");
        assert_eq!(company.name, "テスト株式会社");
        assert_eq!(company.comp_group_code, "GRP001");
    }

    #[test]
    fn test_company_clone() {
        let comp1 = create_test_company();
        let comp2 = comp1.clone();

        assert_eq!(comp1, comp2);
        assert_eq!(comp1.comp_code, comp2.comp_code);
    }

    #[test]
    fn test_company_serialize_deserialize() {
        let company = create_test_company();

        let json = serde_json::to_string(&company).expect("Failed to serialize");
        assert!(!json.is_empty());

        let deserialized: Company = serde_json::from_str(&json).expect("Failed to deserialize");
        assert_eq!(company, deserialized);
    }
}

File: db/scala/src/main/scala/infrastructure/domain/Company.scala
package infrastructure.domain

import scalikejdbc._

import java.time.LocalDateTime

/**
 * 取引先マスタのドメインモデル
 *
 * すべての取引先の基本情報を保持する
 * 顧客、仕入先などの役割に応じた詳細情報は別テーブルで管理（パーティモデル）
 *
 * @param compCode 取引先コード
 * @param name 取引先名
 * @param kana 取引先名カナ
 * @param supType 仕入先区分（0:通常、1:メーカー、2:商社）
 * @param zipCode 郵便番号
 * @param state 都道府県
 * @param address1 住所１
 * @param address2 住所２
 * @param noSalesFlg 取引禁止フラグ（0:許可、1:禁止）
 * @param wideUseType 雑区分
 * @param compGroupCode 取引先グループコード
 * @param maxCredit 与信限度額
 * @param tempCreditUp 与信一時増加枠
 * @param createDate 作成日時
 * @param creator 作成者
 * @param updateDate 更新日時
 * @param updater 更新者
 */
case class Company(
  compCode: String,
  name: String,
  kana: Option[String] = None,
  supType: Int = 0,
  zipCode: Option[String] = None,
  state: Option[String] = None,
  address1: Option[String] = None,
  address2: Option[String] = None,
  noSalesFlg: Int = 0,
  wideUseType: Int = 0,
  compGroupCode: String,
  maxCredit: Int = 0,
  tempCreditUp: Int = 0,
  createDate: LocalDateTime = LocalDateTime.now(),
  creator: String = "",
  updateDate: LocalDateTime = LocalDateTime.now(),
  updater: String = ""
)

object Company extends SQLSyntaxSupport[Company] {
  override val tableName = "取引先マスタ"
  override val columns = Seq(
    "取引先コード", "取引先名", "取引先名カナ", "仕入先区分",
    "郵便番号", "都道府県", "住所１", "住所２",
    "取引禁止フラグ", "雑区分", "取引先グループコード",
    "与信限度額", "与信一時増加枠",
    "作成日時", "作成者名", "更新日時", "更新者名"
  )

  def apply(rs: WrappedResultSet): Company = Company(
    compCode = rs.string("取引先コード"),
    name = rs.string("取引先名"),
    kana = rs.stringOpt("取引先名カナ"),
    supType = rs.int("仕入先区分"),
    zipCode = rs.stringOpt("郵便番号"),
    state = rs.stringOpt("都道府県"),
    address1 = rs.stringOpt("住所１"),
    address2 = rs.stringOpt("住所２"),
    noSalesFlg = rs.int("取引禁止フラグ"),
    wideUseType = rs.int("雑区分"),
    compGroupCode = rs.string("取引先グループコード"),
    maxCredit = rs.int("与信限度額"),
    tempCreditUp = rs.int("与信一時増加枠"),
    createDate = rs.localDateTime("作成日時"),
    creator = rs.string("作成者名"),
    updateDate = rs.localDateTime("更新日時"),
    updater = rs.string("更新者名")
  )
}
